File: src/dynamicTyping/CSharp/duck.cs

interface Duck {
	string quack();
	string dance();
}

class Mallard : Duck {
	public string quack() { return "quack quack"; }
	public string dance() { return " _/¯ "; }	
}

class Wolf {
	public string eat(Duck d) { return " 😈 "; }
	public string quack() { return "QUACK QUACK WHOO"; }
	public string dance() { return " ¯\\_()_/¯ "; }	
}


class Program {
	static void Main(string[] args) {
		Wolf theWolf = new Wolf();
		dynamic theDuck = theWolf; // No necessary.

		Duck aDuck = new Mallard();

		twoDucksAlone(aDuck, theDuck);

		System.Console.WriteLine(theWolf.eat(aDuck));
	}

	public static void twoDucksAlone(dynamic d1, dynamic d2) {
		System.Console.WriteLine(d1.quack());
		System.Console.WriteLine(d2.quack());

		System.Console.WriteLine(d1.dance());
		System.Console.WriteLine(d2.dance());
	}
}

File: src/dynamicTyping/php/duck.php
<?php

interface Duck {
	public function quack() : string;
	public function dance() : string;
}

class Mallard implements Duck {
	public function quack() : string {
		return "quack quack";
	}

	public function dance() : string {
		return " _/¯ ";
	}
}

class Wolf {
	public function quack() : string {
		return "QUACK QUACK WHOO";
	}

	public function dance() : string {
		return " ¯\\_()_/¯ ";
	}

	public function eat(Duck $duck) : string {
		return " 😈 ";
	}
}


// Since php7 variables can be explicitly typed leading to the php nominal typing not to compile the code
// since Wolf is not a Duck. So, weak typing is used here.
function twoDucksAlone($d1, $d2) {
	echo $d1->quack(). "\n";
	echo $d2->quack(). "\n";
	echo $d1->dance(). "\n";
	echo $d2->dance(). "\n";
}


$duck = new Mallard();
$wolf = new Wolf();

twoDucksAlone($duck, $wolf);

echo $wolf->eat($duck) . "\n";

?>

